icefoganalytics/internal-data-portal
# Data Owner: Add Ability To Define Access Constraints

Relates to:
- [0001, Data Modeling, 2024-01-05](https://docs.google.com/document/d/1Mdbl_Cy0BgKiq0DGdvOKUqJn-FdgbF0LSVykwf__FY8/edit?pli=1)
- [0002, UI Modeling, 2024-01-08](https://docs.google.com/document/d/1tP63Ob4XPBGlASPHVf1GnNb3hlIkybmWBqJDA61u3J0/edit)
- https://github.com/icefoganalytics/internal-data-portal/issues/38

# Context

Only building Data Access, and Access panels.

## Dataset Access Page - /datasets/:slug/access

### Front-end routes

Only the page route, though one could argue that the nearby tabs should be included. Note of the components on this page redirect away from it, they just open modals

### Back-end routes

* /api/datasets/:id - GET or maybe /datasets/:slug_or_id
* /api/access-grants?where[data_set_id=... - Get all access grants for a given dataset
* /api/access-grants - POST, create a access grant
* /api/access-grants/:id - PATCH, update an access grant
* /api/access-grants/:id - DELETE, delete an access grant

### Components

* `BaseLayout.vue` - see previous definition
* `DatasetLayout.vue` - see previous definition
* `DatasetAccessPage.vue` - only available to higher rank users
    * `DataAccessFormCard.vue`
        * API URL - Dataset#subscription_url
        * Subscribe Licence - Dataset#subscription_license
    * `AccessCard.vue`
        * `CreateAccessGrantDialog.vue`
            * Shared With - AccessGrant#grant_level
            * Access Type - AccessGrant#access_type
            * Request email - (optional) AccessGrant#requestor_id, User#id, not likely the current user
            * Project Description Required? - AccessGrant#project_description_required
        * `AddAccessGrantButton.vue` - opens access grant creation dialog, or go to page for access grant creation
        * `AccessGrantsEditTable.vue`
            * Standard edit table with update/delete per-record

### Source Wireframe

![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/3df034b2-899f-43b1-8fb2-5a17b057dd80)

# Implementation

Fix modeling around access grant request id; now its a support id and means id of user who responds to access requests, if different from dataset support user.
Add editable Data Access card.
Add editable access grant table, with create/update/delete via modals.
Make user information is always public via policy because the source data is public from YG.

Trigger logout on failed page loads. Not sure if this is the best approach since I didn't really identify why the failure occurred, but it seemed like a better user experience that an infinite page loader.

Render dataset layout tabs dynamically based on user roles.

# Screenshots

![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/7f2ab3a5-33a7-4f7a-b312-125fd48ea014)

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
2. Boot the app via `dev up`
3. Log in to the app at http://localhost:8080
4. From the /dashboard route, either create a new dataset or go to the "view datasets" page and select one from there.
5. After you are viewing a dataset, go to the "Access" tab and check out the new features.
6. Check that you can add/view an api url and subscription license. I'm not sure, but I think the subscription license is an api key?
7. Check that you can add access grants in the "Access" table.
8. Check that you can edit access grants.
9. Check that attempting to duplicate access grants throws a validation error.
10. Check that you can delete access grants.
